zkevm-circuits/src/keccak_circuit/util.rs

```diff
@@ -0,0 +1,320 @@
+        }
+        bytes
+    }
+}
+
+/// Rotates a word that was split into parts to the right
+pub fn rotate<T>(parts: Vec<T>, count: usize, part_size: usize) -> Vec<T> {
+    let mut rotated_parts = parts;
+    rotated_parts.rotate_right(get_rotate_count(count, part_size));
+    rotated_parts
+}
+
+/// Rotates a word that was split into parts to the left
+pub fn rotate_rev<T>(parts: Vec<T>, count: usize, part_size: usize) -> Vec<T> {
+    let mut rotated_parts = parts;
+    rotated_parts.rotate_left(get_rotate_count(count, part_size));
+    rotated_parts
+}
+
+/// Encodes the data using rlc
+pub mod compose_rlc {
+    use eth_types::Field;
+    use halo2_proofs::plonk::Expression;
+
+    pub(crate) fn expr<F: Field>(expressions: &[Expression<F>], r: F) -> Expression<F> {
+        let mut rlc = expressions[0].clone();
+        let mut multiplier = r;
+        for expression in expressions[1..].iter() {
+            rlc = rlc + expression.clone() * multiplier;
+            multiplier *= r;
+        }
+        rlc
+    }
+}
+
+/// Scatters a value into a packed word constant
+pub mod scatter {
+    use super::BIT_SIZE;
+    use eth_types::Field;
+    use halo2_proofs::plonk::Expression;
+
+    pub(crate) fn expr<F: Field>(value: usize, count: usize) -> Expression<F> {
+        let mut packed = F::zero();
+        let mut factor = F::one();
+        for _ in 0..count {
+            packed += F::from(value as u64) * factor;
+            factor *= F::from(BIT_SIZE as u64);
+        }
+        Expression::Constant(packed)
+    }
+}
+
+/// The words that absorb data
+pub fn get_absorb_positions() -> Vec<(usize, usize)> {
+    let mut absorb_positions = Vec::new();
+    for j in 0..5 {
+        for i in 0..5 {
+            if i + j * 5 < 17 {
+                absorb_positions.push((i, j));
+            }
+        }
+    }
+    absorb_positions
+}
+
+/// Converts bytes into bits
+pub fn into_bits(bytes: &[u8]) -> Vec<u8> {
+    let num_bits = bytes.len() * 8;
+    let mut bits: Vec<u8> = vec![0; num_bits];
+
+    let mut counter = 0;
+    for byte in bytes {
+        for idx in 0u64..8 {
+            bits[counter] = (*byte >> idx) & 1;
+            counter += 1;
+        }
+    }
+
+    bits
+}
+
+/// Converts bits into bytes
+pub fn from_bits(bits: &[u8]) -> Word {
+    let mut value = Word::zero();
+    for (idx, bit) in bits.iter().enumerate() {
+        value += Word::from(*bit as u64) << idx;
+    }
+    value
+}
+
+/// Pack bits into a word
+pub fn pack(bits: &[u8]) -> Word {
+    let mut packed = Word::zero();
+    let mut factor = Word::from(1u64);
+    for bit in bits {
+        packed += Word::from(*bit as u64) * factor;
+        factor *= BIT_SIZE;
+    }
+    packed
+}
+
+/// Unpack a word into bits
+pub fn unpack(packed: Word) -> [u8; 64] {
+    let mut bits = [0; 64];
+    for (idx, bit) in bits.iter_mut().enumerate() {
+        *bit = ((packed >> (idx * BIT_COUNT)) & Word::from(BIT_SIZE - 1)).as_u32() as u8;
```

Comment:
Does this "bit" have range 0 to 7? Can we name all of them something like "sparse bit" to distinguish from the binary bit?

Reply:
Bits here also anywhere in the [0,7] range, also added some comments here.

I think pretty much everywhere in the packed implementations "bit" means these sparse bits in the [0,7] range so not sure if calling it something else would be that useful because almost everywhere the new name would be used. Even in the bit based implementation the bits are >1 very often.